// astronomy-bundle/solarEclipse/LocalCircumstances.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var eventCircumstancesCalc_1 = require("./calculations/eventCircumstancesCalc");
var createObservationalCircumstances_1 = __importDefault(require("./createObservationalCircumstances"));
var LocalCircumstances = (function () {
    function LocalCircumstances(besselianElements, location) {
        this.besselianElements = besselianElements;
        this.location = location;
    }
    LocalCircumstances.prototype.getObservationalCircumstancesForC1 = function () {
        var circumstances = (0, eventCircumstancesCalc_1.getTimeLocationCircumstancesC1)(this.besselianElements, this.location);
        return (0, createObservationalCircumstances_1.default)(circumstances);
    };
    LocalCircumstances.prototype.getObservationalCircumstancesForC2 = function () {
        var circumstances = (0, eventCircumstancesCalc_1.getTimeLocationCircumstancesC2)(this.besselianElements, this.location);
        return (0, createObservationalCircumstances_1.default)(circumstances);
    };
    LocalCircumstances.prototype.getObservationalCircumstancesForMaxEclipse = function () {
        var circumstances = (0, eventCircumstancesCalc_1.getTimeLocationCircumstancesMaxEclipse)(this.besselianElements, this.location);
        return (0, createObservationalCircumstances_1.default)(circumstances);
    };
    LocalCircumstances.prototype.getObservationalCircumstancesForC3 = function () {
        var circumstances = (0, eventCircumstancesCalc_1.getTimeLocationCircumstancesC3)(this.besselianElements, this.location);
        return (0, createObservationalCircumstances_1.default)(circumstances);
    };
    LocalCircumstances.prototype.getObservationalCircumstancesForC4 = function () {
        var circumstances = (0, eventCircumstancesCalc_1.getTimeLocationCircumstancesC4)(this.besselianElements, this.location);
        return (0, createObservationalCircumstances_1.default)(circumstances);
    };
    return LocalCircumstances;
}());
exports.default = LocalCircumstances;
