// astronomy-bundle/solarEclipse/ObservationalCircumstances.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var observationalCircumstancesCalc_1 = require("./calculations/observationalCircumstancesCalc");
var ObservationalCircumstances = (function () {
    function ObservationalCircumstances(circumstances) {
        this.circumstances = circumstances;
    }
    ObservationalCircumstances.prototype.getEclipseType = function () {
        return (0, observationalCircumstancesCalc_1.getEclipseType)(this.circumstances);
    };
    ObservationalCircumstances.prototype.getMagnitude = function () {
        return (0, observationalCircumstancesCalc_1.getMagnitude)(this.circumstances);
    };
    ObservationalCircumstances.prototype.getMoonSunRatio = function () {
        return (0, observationalCircumstancesCalc_1.getMoonSunRatio)(this.circumstances);
    };
    ObservationalCircumstances.prototype.getObscuration = function () {
        return (0, observationalCircumstancesCalc_1.getObscuration)(this.circumstances);
    };
    return ObservationalCircumstances;
}());
exports.default = ObservationalCircumstances;
